### PR TITLE
Add service indicator checks and tests for evp ciphers and most AES services

### PR DESCRIPTION
**Pull request by samuel40791765**
### Issues:
Addresses `CryptoAlg-835`

### Description of changes: 
Service indicator checks and tests for:
* AES-ECB:
  * |EVP_CipherFinal_ex|
  * |EVP_Cipher| one-shot function
  * One-shot function: `AES_ecb_encrypt`
* AES-CBC:
  * |EVP_CipherFinal_ex|
  * |EVP_Cipher| one-shot function
  * One-shot function: `AES_cbc_encrypt`
* AES-CTR:
  * |EVP_CipherFinal_ex|
  * |EVP_Cipher| one-shot function
  * One-shot function: `AES_ctr128_encrypt`
* AES-CCM:
  * within underlying aead static functions:  `aead_aes_ccm_seal_scatter` & `aead_aes_ccm_open_gather`
* AES-KW:
  * One-shot functions: `AES_wrap_key` & `AES_unwrap_key` 
* AES-KWP:
  * One-shot functions: `AES_wrap_key_padded` & `AES_unwrap_key_padded`
  
### Call-outs:
* Aside from adding tests for the approved services, an EVP test to check for AES-OFB is non-approved was added. 
* EVP Cipher operation functions: https://commondatastorage.googleapis.com/chromium-boringssl-docs/cipher.h.html#Cipher-operations
  * Added indicator checks in `EVP_EncryptFinal_ex` and `EVP_DecryptFinal_ex` which are used within `EVP_CipherFinal_ex` and by themselves.  
  * Added indicator checks in the `EVP_Cipher` function

### Testing:
New tests in `service_indicator_test.cc` testing dimension.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: crypto/fipsmodule/cipher/e_aes.c [hunk 1]
```diff
@@ -246,6 +246,7 @@ static int aes_cbc_cipher(EVP_CIPHER_CTX *ctx, uint8_t *out, const uint8_t *in,
     CRYPTO_cbc128_decrypt(in, out, len, &dat->ks.ks, ctx->iv, dat->block);
   }
```
Comment (torben-hansen) [inline crypto/fipsmodule/cipher/e_aes.c:248]: Why add service indicators in this file? These functions are only accessible through high-level functions, no?

Comment (samuel40791765) [reply crypto/fipsmodule/cipher/e_aes.c]: Moved the check to `int EVP_Cipher(EVP_CIPHER_CTX *ctx, uint8_t *out, const uint8_t *in, size_t in_len)` which directly accessed these static functions.

File: crypto/fipsmodule/service_indicator/service_indicator.c [hunk 2]
```diff
@@ -92,6 +113,7 @@ uint64_t FIPS_service_indicator_after_call(void) { return 0; }
 int FIPS_service_indicator_check_approved(int before, int after) { return AWSLC_NOT_APPROVED; }
 
 void AES_verify_service_indicator(unsigned key_rounds) { }
+void AES_verify_service_indicator_with_ctx(EVP_CIPHER_CTX *ctx) { }
```
Comment (darylmartin100) [inline crypto/fipsmodule/service_indicator/service_indicator.c:116]: I like this approach a lot, let the compiler optimize all these functions away when not in FIPS mode.  This should cause unused argument warnings though, I would say either suppress them (#pragma) or do a a void cast on the argument so we can build with -Werror.  I'd also add a sentence or two in code describing why these are empty functions.

File: crypto/fipsmodule/service_indicator/service_indicator.c [hunk 3]
```diff
@@ -92,6 +113,7 @@ uint64_t FIPS_service_indicator_after_call(void) { return 0; }
 int FIPS_service_indicator_check_approved(int before, int after) { return AWSLC_NOT_APPROVED; }
```
Comment (darylmartin100) [inline crypto/fipsmodule/service_indicator/service_indicator.c:113]: If we are not in FIPS mode should we be returning approved for everything instead?  Wouldn't want to impact a random application with incorrect FIPS checks in their code..

File: crypto/fipsmodule/service_indicator/service_indicator.c (hunk 2 above)
Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator.c:116]: I didn't know that void casts were doable in this situation. Added :)

File: crypto/fipsmodule/service_indicator/service_indicator.c (hunk 3 above)
Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator.c:113]: The current approach we've taken this is returning `AWSLC_APPROVED` for the wrapper macro we're advocating that people use, and `AWSLC_NOT_APPROVED` for the actual approval check. I do see how this might be a bit confusing however.. Should I change the function to align with the macro?
Macro: https://github.com/awslabs/aws-lc/blob/main/include/openssl/service_indicator.h#L52-L59
Current Test: https://github.com/awslabs/aws-lc/blob/main/crypto/fipsmodule/service_indicator/service_indicator_test.cc#L205-L219

Comment (darylmartin100) [reply crypto/fipsmodule/service_indicator/service_indicator.c]: I'm fine either way as long as it's documented well enough so I can choose my own adventure to find the right answer :)  Adding a few comments would make me happy.

Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator.c]: Added description in `service_indicator.h`.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 4]
```diff
@@ -130,30 +305,87 @@ TEST(ServiceIndicatorTest, BasicTest) {
...
   ASSERT_EQ(approved, AWSLC_NOT_APPROVED);
 }
 
+TEST(ServiceIndicatorTest, AESECB) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  AES_KEY aes_key;
+  uint8_t output[256];
+
+  // AES-CBC Encryption KAT
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:319]: AES-ECB?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 5]
```diff
@@ -130,30 +305,87 @@ TEST(ServiceIndicatorTest, BasicTest) {
...
+TEST(ServiceIndicatorTest, AESCTR) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  AES_KEY aes_key;
+  uint8_t aes_iv[16];
+  uint8_t output[256];
+  unsigned num = 0;
+  uint8_t ecount_buf[AES_BLOCK_SIZE];
+
+  // AES-CBC Encryption KAT
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:374]: AES-CTR?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 6]
```diff
@@ -189,6 +421,82 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  ASSERT_TRUE(EVP_AEAD_CTX_init(aead_ctx.get(), EVP_aead_aes_128_ccm_bluetooth(),
+                                kAESKey, sizeof(kAESKey), EVP_AEAD_DEFAULT_TAG_LENGTH, nullptr));
+
+  // AES-CCM Encryption
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,EVP_AEAD_CTX_seal(aead_ctx.get(),
+       output, &out_len, sizeof(output), nonce, EVP_AEAD_nonce_length(EVP_aead_aes_128_ccm_bluetooth()),
+       kPlaintext, sizeof(kPlaintext), nullptr, 0));
+  ASSERT_TRUE(check_test(kAESCCMCiphertext, output, sizeof(kAESCCMCiphertext),
+                         "AES-CCM Encryption KAT"));
+  ASSERT_TRUE(approved);
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:442]: Same comment here as above about using `ASSERT_EQ` and the occurrence below.

File: crypto/fipsmodule/cipher/cipher.c [hunk 7]
```diff
@@ -225,6 +225,7 @@ int EVP_CipherInit_ex(EVP_CIPHER_CTX *ctx, const EVP_CIPHER *cipher,
 
   ctx->buf_len = 0;
   ctx->final_used = 0;
+  AES_verify_service_indicator_with_ctx(ctx);
```
Comment (nebeid) [inline crypto/fipsmodule/cipher/cipher.c:228]: Thanks for taking care of every successful return path, Samuel. Do you know if they were all covered by the tests?

File: crypto/fipsmodule/service_indicator/service_indicator.c [hunk 8]
```diff
@@ -85,14 +106,43 @@ void AEAD_verify_service_indicator(size_t key_length) {
...
+int FIPS_service_indicator_check_approved(int before, int after) {
+  (void)before;
+  (void)after;
+  return AWSLC_NOT_APPROVED;
+}
+
+
+// Service indicator check functions listed below are optimized to not do extra
+// checks, when not in FIPS mode. Arguments must be cast to void to avoid
+// compiler errors.
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator.c:129]: Should this comment precede the function above?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 9]
```diff
@@ -34,6 +35,93 @@ static const uint8_t kPlaintext[64] = {
...
+  }
+  std::vector<uint8_t> iv(kAESIV,kAESIV + EVP_CIPHER_CTX_iv_length(ctx1.get()));
+  ASSERT_EQ(iv.size(), EVP_CIPHER_CTX_iv_length(ctx1.get()));
+
+
+  bssl::ScopedEVP_CIPHER_CTX ctx2;
+  ASSERT_TRUE(EVP_CIPHER_CTX_copy(ctx2.get(), ctx1.get()));
+  EVP_CIPHER_CTX *ctx = ctx2.get();
+
+  // The ciphers are run with no padding. For each of the ciphers we test, the
```
[the remark this answers is not in the document]
Comment (samuel40791765) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc]: I took this from the original `cipher` tests and didn't give much thought to this part. They were probably trying to take care of other cases outside of this scope. Removed it.

File: crypto/fipsmodule/cipher/cipher.c [hunk 10]
```diff
@@ -480,12 +493,17 @@ int EVP_DecryptFinal_ex(EVP_CIPHER_CTX *ctx, unsigned char *out, int *out_len) {
...
 
+  AES_verify_service_indicator_with_ctx(ctx);
   return 1;
 }
 
 int EVP_Cipher(EVP_CIPHER_CTX *ctx, uint8_t *out, const uint8_t *in,
                size_t in_len) {
-  return ctx->cipher->cipher(ctx, out, in, in_len);
+  int ret = ctx->cipher->cipher(ctx, out, in, in_len);
+  if(ret == 1) {
```
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/fipsmodule/cipher/cipher.c:507]: > For stream ciphers, the stream is advanced past the bytes
// used. It returns one on success and zero otherwise, unless |EVP_CIPHER_flags|
// has |EVP_CIPH_FLAG_CUSTOM_CIPHER| set. Then it returns the number of bytes
// written or -1 on error.

I guess the more precise statement is that the return value depends on the actual cipher... Since you don't know that here, well, I don't think you can just check `== 1`.

You can make a bigger argument that we only validate ciphers that doesn't set `EVP_CIPH_FLAG_CUSTOM_CIPHER`.  But there is no need to make that argument imo.

File: include/openssl/service_indicator.h [hunk 11]
```diff
@@ -30,9 +30,9 @@ OPENSSL_EXPORT uint64_t FIPS_service_indicator_after_call(void);
 //
 // Service indicator calls should not be used in non-FIPS builds. However, if
 // used, the direct check to |FIPS_service_indicator_check_approved|
-// will always indicate |AWSLC_NOT_APPROVED| during non-FIPS.
-// It is recommended to use the macro |CALL_SERVICE_AND_CHECK_APPROVED|, which
-// will always return |AWSLC_APPROVED| during non-FIPS.
+// will always indicate |AWSLC_APPROVED| during non-FIPS.
+// It is recommended to use the macro |CALL_SERVICE_AND_CHECK_APPROVED| though,
+// which will also always return |AWSLC_APPROVED| during non-FIPS.
```
[the remark this answers is not in the document]
Comment (nebeid) [reply include/openssl/service_indicator.h:35]: You can ignore this suggestion since you started using "(non-)FIPS mode" in PR#263

File: crypto/fipsmodule/service_indicator/service_indicator.c [hunk 12]
```diff
@@ -59,40 +60,78 @@ void FIPS_service_indicator_update_state(void) {
...
-int FIPS_service_indicator_check_approved(int before, int after) { return AWSLC_NOT_APPROVED; }
 
-void AES_verify_service_indicator(unsigned key_rounds) { }
-void AEAD_verify_service_indicator(size_t key_length) { }
+
+// Service indicator check functions listed below are optimized to not do extra
+// checks, when not in FIPS mode. Arguments must be cast to void to avoid
+// unused warnings.
+
+int FIPS_service_indicator_check_approved(OPENSSL_UNUSED int before,
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator.c:124]: In #263 https://github.com/awslabs/aws-lc/pull/263/files#diff-ce7ffb6368fc741735e07a58ef82c4e0bb9b31d389299f8d0d1e961c76dd9f38R147

this returns `AWSLC_NOT_APPROVED`.

File: crypto/fipsmodule/service_indicator/internal.h [hunk 13]
```diff
@@ -32,9 +33,16 @@ OPENSSL_INLINE void FIPS_service_indicator_update_state(void) { }
...
+// available to check against.
+void AES_verify_service_indicator_with_ctx(EVP_CIPHER_CTX *ctx);
+
 // The service indicator checks use different parameters for AEAD APIs than
 // those of other AES modes. AES-GCM is approved only with an internal IV, see
 // SP 800-38D Sec 8.2.2.
 void AEAD_verify_service_indicator(size_t key_length);
 
+// Only 128 bit keys with 32 bit tag lengths are approved for AES-CCM.
+void AEAD_CCM_verify_service_indicator(const EVP_AEAD_CTX *ctx);
```
[the remark this answers is not in the document]
Comment (nebeid) [reply crypto/fipsmodule/service_indicator/internal.h:44]: Before I forget, should we add tests for GCM-256 being approved and GCM-192 not approved?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 14]
```diff
@@ -1,15 +1,14 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc:1]: Unit test: Add `EVP_{Encrypt,Decrypt}Final_ex` for completeness
Unit test: can we use `3des` with EVP_Cipher APIs? If yes, add test that SI is not flipped to `approved` state

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 15]
```diff
@@ -58,6 +63,93 @@ static int check_test(const uint8_t *expected, const uint8_t *actual,
...
+  size_t max_out = in.size();
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
+    unsigned block_size = EVP_CIPHER_CTX_block_size(ctx);
+    max_out += block_size - (max_out % block_size);
+  }
+  out->resize(max_out);
+
+  size_t total = 0;
+  int len;
+  EXPECT_TRUE(EVP_CipherUpdate(ctx, out->data(), &len, in.data(), in.size()));
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:78]: `ASSERT_TRUE` or `EXPECT_TRUE`?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 16]
```diff
@@ -58,6 +63,93 @@ static int check_test(const uint8_t *expected, const uint8_t *actual,
...
+  total += static_cast<size_t>(len);
+  // Check if the overall service is approved by checking |EVP_CipherFinal_ex|,
+  // which should be the last part of the service.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,EXPECT_TRUE(
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:82]: Same.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 17]
```diff
@@ -58,6 +63,93 @@ static int check_test(const uint8_t *expected, const uint8_t *actual,
...
+    in = &ciphertext;
+    out = &plaintext;
+  }
+
+  bssl::ScopedEVP_CIPHER_CTX ctx;
+  // Test running the EVP_Cipher interfaces one by one directly, and check
+  // |EVP_CipherFinal_ex| for approval at the end.
+  ASSERT_TRUE(EVP_CipherInit_ex(ctx.get(), cipher, nullptr, nullptr,
+                                    nullptr, encrypt ? 1 : 0));
+  std::vector<uint8_t> iv( kAESIV,kAESIV + EVP_CIPHER_CTX_iv_length(ctx.get()));
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:110]: Nit: some random spacing in this line.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 18]
```diff
@@ -58,6 +63,93 @@ static int check_test(const uint8_t *expected, const uint8_t *actual,
...
+  ASSERT_EQ(iv.size(), EVP_CIPHER_CTX_iv_length(ctx.get()));
+
+  // For each of the ciphers we test, the output size matches the input size.
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:113]: Is this relevant to the SI?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 19]
```diff
@@ -157,6 +288,35 @@ static const uint8_t kHMACOutput_sha512_256[SHA512_256_DIGEST_LENGTH] = {
...
+  { EVP_aes_128_ctr(), kAESCTRCiphertext, 64, true, AWSLC_APPROVED },
+  { EVP_aes_128_ofb(), kAESOFBCiphertext, 64, true, AWSLC_NOT_APPROVED }
+};
+
+class EVP_ServiceIndicatorTest : public testing::TestWithParam<EVP_TestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, EVP_ServiceIndicatorTest, testing::ValuesIn(nTestVectors));
+
+TEST_P(EVP_ServiceIndicatorTest, EVP_Ciphers) {
+  const EVP_TestVector &t = GetParam();
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:309]: niiiit: Use more descriptive name; as here https://github.com/awslabs/aws-lc/commit/c5986770044dbfa420a1aa4317195aa23be73b90#diff-ff3c1dd2d1224704ce9f4249523fbd553b13e40e9431607f3ee5489af2298358R203

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc:1]: Please fix the text fixture name for this
```
TEST_P(EVP_MD_ServiceIndicatorTest, EVP_Ciphers) {
```
Does all `TEST_P` type tests have to have distinct group? If no, maybe we should just have the same group

If fixing: Maybe better to do a quick PR after this one that consolidates the grouping.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 20]
```diff
@@ -377,34 +537,89 @@ TEST(ServiceIndicatorTest, BasicTest) {
...
+
 TEST(ServiceIndicatorTest, AESCBC) {
   int approved = AWSLC_NOT_APPROVED;
   AES_KEY aes_key;
   uint8_t aes_iv[16];
   uint8_t output[256];
   // AES-CBC Encryption KAT
   memcpy(aes_iv, kAESIV, sizeof(kAESIV));
-  ASSERT_TRUE(AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key) == 0);
+  ASSERT_EQ(AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key), 0);
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:582]: Why change this?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 21]
```diff
@@ -377,34 +537,89 @@ TEST(ServiceIndicatorTest, BasicTest) {
...
+}
+
+TEST(ServiceIndicatorTest, AESCTR) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  AES_KEY aes_key;
+  uint8_t aes_iv[16];
+  uint8_t output[256];
+  unsigned num = 0;
+  uint8_t ecount_buf[AES_BLOCK_SIZE];
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:606]: `ecount`? enlighten me :)

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (torben-hansen) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc:1]: We should have some basic tests that doesn't use the macro - add in a separate PR

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 22]
```diff
@@ -438,6 +653,86 @@ TEST(ServiceIndicatorTest, AESGCM) {
       encrypt_output, &out_len, sizeof(encrypt_output), nonce, EVP_AEAD_nonce_length(EVP_aead_aes_128_gcm()),
           kPlaintext, sizeof(kPlaintext), nullptr, 0));
   ASSERT_EQ(approved, AWSLC_NOT_APPROVED);
+
+  // Call non-approved external IV usage of AES-GCM 192 bit key size. (192 is
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:657]: Also add test for the regular gcm algorithms that are not approved, not only the 192-bit one?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 23]
```diff
@@ -438,6 +653,86 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  // not available for internal IV.)
+  ASSERT_TRUE(EVP_AEAD_CTX_init(aead_ctx.get(), EVP_aead_aes_192_gcm(),
+                              kAESKey_192, sizeof(kAESKey_192), 0, nullptr));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, EVP_AEAD_CTX_seal(aead_ctx.get(),
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:661]: Verify return value

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 24]
```diff
@@ -438,6 +653,86 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  uint8_t nonce[EVP_AEAD_MAX_NONCE_LENGTH];
+  uint8_t output[256];
+  size_t out_len;
+
+  OPENSSL_memset(nonce, 0, sizeof(nonce));
+  ASSERT_TRUE(EVP_AEAD_CTX_init(aead_ctx.get(), EVP_aead_aes_128_ccm_bluetooth(),
+                                kAESKey, sizeof(kAESKey), EVP_AEAD_DEFAULT_TAG_LENGTH, nullptr));
+
+  // AES-CCM Encryption
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,EVP_AEAD_CTX_seal(aead_ctx.get(),
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:680]: Verify return value

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 25]
```diff
@@ -438,6 +653,86 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+       output, &out_len, sizeof(output), nonce, EVP_AEAD_nonce_length(EVP_aead_aes_128_ccm_bluetooth()),
+       kPlaintext, sizeof(kPlaintext), nullptr, 0));
+  ASSERT_TRUE(check_test(kAESCCMCiphertext, output, out_len, "AES-CCM Encryption KAT"));
+  ASSERT_EQ(approved, AWSLC_APPROVED);
+
+  // AES-CCM Decryption
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,EVP_AEAD_CTX_open(aead_ctx.get(),
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:687]: Verify return value

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 26]
```diff
@@ -438,6 +653,86 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+TEST(ServiceIndicatorTest, AESKW) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  AES_KEY aes_key;
+  uint8_t output[256];
+  size_t outlen;
+
+  // AES-KW Encryption KAT
+  ASSERT_EQ(AES_set_encrypt_key(kAESKey, 8 * sizeof(kAESKey), &aes_key),0);
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,outlen = AES_wrap_key(&aes_key, nullptr,
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:703]: Verify return value `outlen` is as expected? (also spacing)

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 27]
```diff
@@ -464,10 +758,10 @@ TEST(ServiceIndicatorTest, BasicTest) {
   // Macro should return true, to ensure FIPS/Non-FIPS compatibility.
   ASSERT_EQ(approved, AWSLC_APPROVED);
 
-  // Actual approval check should return false during non-FIPS.
+  // Actual approval check should also return true during non-FIPS.
```
Comment (torben-hansen) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:761]:
```suggestion
  // Approval check should also return true when not in FIPS mode.
```

File: crypto/fipsmodule/cipher/cipher.c [hunk 28]
```diff
@@ -439,8 +442,7 @@ int EVP_DecryptFinal_ex(EVP_CIPHER_CTX *ctx, unsigned char *out, int *out_len) {
     } else {
       *out_len = i;
     }
-    AES_verify_service_indicator(ctx, 0);
-    return 1;
+    goto success;
```
Comment (nebeid) [inline crypto/fipsmodule/cipher/cipher.c:445]: The destination of goto in EVP_DecryptFinal_ex is always success, which is different from EVP_EncryptFinal_ex, is that intended? Is there nothing that may need locking in the call on l.439?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 29]
```diff
@@ -417,27 +717,134 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  ASSERT_EQ(approved, AWSLC_APPROVED);
+
+  // Call approved internal IV usage of AES-GCM 256 bit kye size.
+  ASSERT_TRUE(EVP_AEAD_CTX_init(aead_ctx.get(), EVP_aead_aes_256_gcm_randnonce(),
+                                kAESKey_256, sizeof(kAESKey_256), 0, nullptr));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,ASSERT_TRUE(EVP_AEAD_CTX_seal(aead_ctx.get(),
+      encrypt_output, &out_len, sizeof(encrypt_output), nullptr, 0, kPlaintext, sizeof(kPlaintext), nullptr, 0)));
   ASSERT_EQ(approved, AWSLC_APPROVED);
 
-  CALL_SERVICE_AND_CHECK_APPROVED(approved,EVP_AEAD_CTX_open(aead_ctx.get(),
```
Comment (samuel40791765) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:428]: spacing

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 30]
```diff
@@ -417,27 +717,134 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  uint8_t nonce[EVP_AEAD_MAX_NONCE_LENGTH];
+  uint8_t output[256];
+  size_t out_len;
+
+  OPENSSL_memset(nonce, 0, sizeof(nonce));
+  ASSERT_TRUE(EVP_AEAD_CTX_init(aead_ctx.get(), EVP_aead_aes_128_ccm_bluetooth(),
+                                kAESKey, sizeof(kAESKey), EVP_AEAD_DEFAULT_TAG_LENGTH, nullptr));
+
+  // AES-CCM Encryption
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,ASSERT_TRUE(EVP_AEAD_CTX_seal(aead_ctx.get(),
```
Comment (samuel40791765) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:790]: spacing

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 31]
```diff
@@ -192,42 +227,292 @@ static const uint8_t kAESKWPCiphertext[72] = {
...
+
+TEST(ServiceIndicatorTest, CMAC) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  std::vector<uint8_t> mac(16);
+  size_t out_len;
+  bssl::UniquePtr<CMAC_CTX> ctx(CMAC_CTX_new());
+  ASSERT_TRUE(ctx);
+
+  // Test running the CMAC interfaces one by one directly, and check
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc]: I see there is only one interface, so this comment needs adjustment.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 32]
```diff
@@ -382,26 +717,60 @@ TEST(ServiceIndicatorTest, AESGCM) {
...
+  // Call approved internal IV usage of AES-GCM 256 bit kye size.
+  ASSERT_TRUE(EVP_AEAD_CTX_init(aead_ctx.get(), EVP_aead_aes_256_gcm_randnonce(),
+                                kAESKey_256, sizeof(kAESKey_256), 0, nullptr));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,ASSERT_TRUE(EVP_AEAD_CTX_seal(aead_ctx.get(),
+      encrypt_output, &out_len, sizeof(encrypt_output), nullptr, 0, kPlaintext, sizeof(kPlaintext), nullptr, 0)));
   ASSERT_EQ(approved, AWSLC_APPROVED);
 
-  CALL_SERVICE_AND_CHECK_APPROVED(approved,EVP_AEAD_CTX_open(aead_ctx.get(),
-      decrypt_output, &out2_len, sizeof(decrypt_output), nullptr, 0, encrypt_output, out_len, nullptr, 0));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,ASSERT_TRUE(EVP_AEAD_CTX_open(aead_ctx.get(),
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc]:
```suggestion
  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_AEAD_CTX_open(aead_ctx.get(),
```

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 33]
```diff
@@ -418,16 +787,16 @@ TEST(ServiceIndicatorTest, AESCCM) {
...
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,ASSERT_TRUE(EVP_AEAD_CTX_seal(aead_ctx.get(),
        output, &out_len, sizeof(output), nonce, EVP_AEAD_nonce_length(EVP_aead_aes_128_ccm_bluetooth()),
-       kPlaintext, sizeof(kPlaintext), nullptr, 0));
+       kPlaintext, sizeof(kPlaintext), nullptr, 0)));
   ASSERT_TRUE(check_test(kAESCCMCiphertext, output, out_len, "AES-CCM Encryption KAT"));
   ASSERT_EQ(approved, AWSLC_APPROVED);
 
   // AES-CCM Decryption
-  CALL_SERVICE_AND_CHECK_APPROVED(approved,EVP_AEAD_CTX_open(aead_ctx.get(),
+  CALL_SERVICE_AND_CHECK_APPROVED(approved,ASSERT_TRUE(EVP_AEAD_CTX_open(aead_ctx.get(),
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc]:
```suggestion
  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(EVP_AEAD_CTX_open(aead_ctx.get(),
```

**Review by nebeid — COMMENTED** [review]
will check the next 2 commits, may be these comments are already resolved.

File: crypto/fipsmodule/cipher/cipher.c [hunk 34]
```diff
@@ -322,15 +322,15 @@ int EVP_EncryptUpdate(EVP_CIPHER_CTX *ctx, uint8_t *out, int *out_len,
...
 int EVP_EncryptFinal_ex(EVP_CIPHER_CTX *ctx, uint8_t *out, int *out_len) {
-  int n, ret = 0;
+  int n, ret = 0, custom_ret;
   unsigned int i, b, bl;
 
   if (ctx->cipher->flags & EVP_CIPH_FLAG_CUSTOM_CIPHER) {
-    ret = ctx->cipher->cipher(ctx, out, NULL, 0);
-    if (ret < 0) {
-      return 0;
+    custom_ret = ctx->cipher->cipher(ctx, out, NULL, 0);
```
Comment (nebeid) [inline crypto/fipsmodule/cipher/cipher.c:329]: You may want to add a comment that this `cipher()` pointer is always a static function which won't contain Service Indicator updates (if I understood our discussion correctly). I see you added that comment in DecryptFinal. I think it's needed in EncryptFinal as well, beside here.

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 35]
```diff
@@ -82,6 +82,32 @@ static void DoCipher(EVP_CIPHER_CTX *ctx, std::vector<uint8_t> *out,
   int len;
   ASSERT_TRUE(EVP_CipherUpdate(ctx, out->data(), &len, in.data(), in.size()));
   total += static_cast<size_t>(len);
+  // Check if the overall service is approved by checking |EVP_EncryptFinal_ex|
+  // or |EVP_DecryptFinal_ex|, which should be the last part of the service.
+  if (ctx->encrypt) {
+    CALL_SERVICE_AND_CHECK_APPROVED(approved, EVP_EncryptFinal_ex(ctx, out->data() + total, &len));
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc:88]: Should you be adding ASSERT_TRUE on the function call itself as elsewhere?

File: crypto/fipsmodule/service_indicator/service_indicator_test.cc [hunk 36]
```diff
@@ -108,7 +134,7 @@ static void TestOperation(const EVP_CIPHER *cipher, bool encrypt,
 
   bssl::ScopedEVP_CIPHER_CTX ctx;
   // Test running the EVP_Cipher interfaces one by one directly, and check
-  // |EVP_CipherFinal_ex| for approval at the end.
```
Comment (nebeid) [inline crypto/fipsmodule/service_indicator/service_indicator_test.cc]: You also do check `EVP_CipherFinal_ex` as well, correct?

Comment (nebeid) [reply crypto/fipsmodule/service_indicator/service_indicator_test.cc]: See https://github.com/awslabs/aws-lc/blob/af590a7fa59d1d2b6e30df2e4b3aec7dc0c08957/crypto/fipsmodule/service_indicator/service_indicator_test.cc#L150-162. Resolving.

**Review by nebeid — APPROVED** [review]
